### PR TITLE
feat(alerting): add query optimizations for prometheus

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
This PR introduces the same set of optimizations we already did for Loki. If applicable it will transform a `range` query to and `instant` query.

### CODE REVIEW

File: pkg/services/ngalert/store/range_to_instant_test.go [hunk 1]
```diff
@@ -199,6 +312,27 @@ func lokiQuery(t *testing.T, refID, queryType, datasourceUID string) models.Aler
 	}
 }
 
+func promethesQuery(t *testing.T, refID, datasourceUID string, isInstant bool) models.AlertQuery {
```
Comment (stevesg) [inline pkg/services/ngalert/store/range_to_instant_test.go:315]:
```suggestion
func prometheusQuery(t *testing.T, refID, datasourceUID string, isInstant bool) models.AlertQuery {
```

File: pkg/services/ngalert/store/range_to_instant_test.go [hunk 2]
```diff
@@ -178,6 +260,37 @@ func createMultiQueryMigratableLokiRule(t *testing.T, muts ...func(*models.Alert
...
+	}
+	return r
+}
+
+func createMultiQueryMigratablePromRule(t *testing.T, muts ...func(*models.AlertRule)) *models.AlertRule {
+	t.Helper()
+	r := &models.AlertRule{
+		Data: []models.AlertQuery{
+			promethesQuery(t, "TotalRequests", "grafanacloud-prom", promIsNotInstant),
+			promethesQuery(t, "TotalErrors", "grafanacloud-prom", promIsNotInstant),
```
Comment (stevesg) [inline pkg/services/ngalert/store/range_to_instant_test.go:283]:
```suggestion
			prometheusQuery(t, "TotalRequests", "grafanacloud-prom", promIsNotInstant),
			prometheusQuery(t, "TotalErrors", "grafanacloud-prom", promIsNotInstant),
```

File: pkg/services/ngalert/store/range_to_instant.go [hunk 3]
```diff
@@ -13,34 +14,48 @@ type dsType struct {
...
 		_ = json.Unmarshal(r.Data[i].Model, &t)
 
-		if !t.isLoki() {
+		if !t.isLoki() && !t.isPrometheus() {
 			continue
 		}
+
+		if (t.isLoki() && r.Data[i].QueryType != "range") || (t.isPrometheus() && !t.Range) {
+			continue
+		}
```
Comment (stevesg) [inline pkg/services/ngalert/store/range_to_instant.go:57]: [nit] Stylistic comment - I find it a little jarring that we use `switch` below, but these `isXXX` functions here. Might look nicer to use a `switch` here too.

**Review by stevesg — APPROVED** [review]
Nice, just nits.

Just one question - did you consider a feature flag for this? I know we started with a feature flag for Loki, but given the similarity, I imagine we have gained enough confidence to not do it here.

File: pkg/services/ngalert/store/range_to_instant.go (hunk 3 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/store/range_to_instant.go]: SOLD

**Comment by JohnnyQQQQ** [comment]
> Nice, just nits.
> 
> Just one question - did you consider a feature flag for this? I know we started with a feature flag for Loki, but given the similarity, I imagine we have gained enough confidence to not do it here.

Yeah, I did consider a feature flag, but as you pointed out we already had the same approach prior behind a feature flag, so I think we are good.

**Pull request merged**